python/external_config/external_config_loader.py

```diff
@@ -0,0 +1,283 @@
+        self._config_state.state_changed.connect(self.configurations_changed.emit)
+        # always trigger a check at startup
+        self.refresh_shotgun_global_state()
+
+        self._bg_task_manager = bg_task_manager
+        self._bg_task_manager.task_completed.connect(self._task_completed)
+        self._bg_task_manager.task_failed.connect(self._task_failed)
+
+    def __repr__(self):
+        """
+        String representation
+        """
+        return "<CommandHandler %s@%s>" % (self._engine_name, self._plugin_id)
+
+    def shut_down(self):
+        """
+        Shut down and deallocate.
+        """
+        self._config_state.shut_down()
+
+    def refresh_shotgun_global_state(self):
+        """
+        Requests an async refresh. If the State of Shotgun has
+        changed in a way which may affect configurations, this will
+        result in a ``configurations_changed`` signal being emitted.
+
+        Examples of state changes which may affect configurations are any changes
+        to related pipeline configuration, but also indirect changes such as a
+        change to the list of software entities, since these can implicitly affect
+        the list of commands associated with a project or entity.
+        """
+        self._config_state.refresh()
+
+    @property
+    def engine_name(self):
+        """
+        The name of the engine associated with this external configuration loader.
+        """
+        return self._engine_name
+
+    @property
+    def interpreter(self):
+        """
+        The Python interpreter to when bootstrapping and loading external configurations.
+        """
+        return self._interpreter
+
+    @property
+    def plugin_id(self):
+        """
+        The plugin id which will be used when executing external requests.
+        """
+        return self._plugin_id
+
+    @property
+    def base_config_uri(self):
+        """
+        Cnfiguration URI string to be used when nothing is provided via Shotgun overrides.
+        """
+        return self._base_config_uri
+
+    def request_configurations(self, project_id, force_reload=False):
+        """
+        Requests a list of configuration objects for the given project.
+
+        Emits a ``configurations_loaded`` signal when the configurations
+        have been loaded.
+
+        :param project_id: Project to request configurations for.
+        :param force_reload: If ``True``, force reload the configuration data.
+            If ``False`` (default), use a cached representation. This
+            cache is refreshed at startup and whenever
+            :meth:`refresh_shotgun_global_state` is called.
+        """
+        # load existing cache file if it exists
+        config_cache_key = {
+            "project": project_id,
+            "plugin": self._plugin_id,
+            "state_hash": self._config_state.get_hash()
+        }
+
+        config_data = file_cache.load_cache(config_cache_key)
```

Comment:
I thought I mentioned it during my first pass, but couldn't retrieve my comment: it seems we're loading the cache for nothing if `force_reload` is `False`?

Reply:
Actually, I was writing a comment about this method being confusing, and I think I understand my confusion: `force_reload` is actually a way to force a refresh, and ignore any cached data. I wrongly associated `load` with loading the cached data, not sending a refresh request. So may be this param should be `force_refresh`, or the docstring be updated with something like:
```
       :param force_reload: If ``True``, force a refresh of the configuration data.
             If ``False`` (default), use a cached representation if available. This
             cache is refreshed at startup and whenever
             :meth:`refresh_shotgun_global_state` is called.
```
It took me a while to get it, so I think it's worth clarifying it. To me `reload` means load the cached data again.